design-proposals/cluster-orch-integrating-k8s-into-emt.md

```diff
@@ -0,0 +1,302 @@
+
+### Workflow Change
+
+Clusters for a specific host can be created in two ways:
+
+**1. Automatically during host registration**: This new option is designed for users with simple use cases who want Kubernetes to be available with a pre-defined template, reducing user interactions during onboarding and cluster creation.
+
+**2. Manually through a direct request to the Cluster Manager**: This method allows users to create clusters on hosts that have already been onboarded (or provisioned, depending on the finalized design). It is particularly useful in scenarios where host onboarding and workload creation occur independently, often managed by different users with distinct roles or at different times. For example, an Onboarding Manager might register a pool of hosts with only the auto-onboard option enabled (and potentially auto-provision, based on the final design). Later, Edge Operators or Users can select an operating system and create clusters using an OS profile, Kubernetes flavor, and template that best align with specific workload requirements. This approach provides flexibility and adaptability, making it well-suited for cloud-like edge usage.
+
+Before diving into workflow changes, it is essential to understand the coupling introduced by this design change between the following components:
+
+- **EMT Image Version**: Specifies the operating system and embedded software, including the K3s version, on the edge device.
+- **Cluster Template**: Defines the configuration for Kubernetes cluster creation, including the Kubernetes flavor, version, and custom control plane settings.
+
+For instance, attempting to create a K3s cluster using a template for K3s v1.32 on an EMT machine embedding K3s v1.30 may result in compatibility issues. Such mismatches can occur when Edge Orchestration undergoes multiple upgrades, leading to a mix of cluster templates and EMT machines with varying versions. While this issue is specific to the K3s-on-EMT scenario, it is critical to address it effectively in the user workflow, as this represents the most common use case for EMF.
+
+#### Automatic Cluster Creation
+
+The first option introduces a streamlined process for users seeking simplicity or bulk registration of hosts, or both. Both the Web UI and CLI will offer a new toggle for `Create Cluster Automatically` (potentially combined with `Provision Automatically`, pending final design decisions) during registration. This option is only valid when `Provision Automatically` is enabled. When enabled, it automatically creates a single-node cluster for each host.
+
+Users can provide a default cluster template for all hosts in the registration request, with the flexibility to override it for specific hosts if needed. For EMT machines, once a specific OS profile is selected, only cluster templates compatible with that EMT version will be displayed. This includes all RKE2 templates and K3s templates that match the K3s version embedded in the selected EMT image. 
```

Comment:
Ho do you override the template to a specific one ? can we have a drop down in the registration instead of a toggle ? where the dropdown allows you to pick the template between any k8s type ?

Reply:
That is more about UX design. Patricia has initial design so feel free to reach out her, but "toggle" is for user to select auto cluster creation enabled or disabled. User should provide more information when enabled in subsequent steps of registration.